### PR TITLE
Move update and delete by query to use seq# for optimistic concurrency control

### PR DESCRIPTION
**Pull request by bleskes**
The delete and update by query APIs both offer protection against overriding concurrent user changes to the documents they touch. They currently are using internal versioning. This PR changes that to rely on sequences numbers and primary terms.

Relates #37639 
Relates #36148 
Relates #10708

### CODE REVIEW

**Review by ywelsch — CHANGES_REQUESTED** [review]
One nit and possibly one important comment

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/AsyncDeleteByQueryAction.java [hunk 1]
```diff
@@ -31,20 +32,20 @@
  * Implementation of delete-by-query using scrolling and bulk.
  */
 public class AsyncDeleteByQueryAction extends AbstractAsyncBulkByScrollAction<DeleteByQueryRequest> {
+    private final boolean useSeqNoForOCC;
```
[the remark this answers is not in the document]
Comment (bleskes) [reply modules/reindex/src/main/java/org/elasticsearch/index/reindex/AsyncDeleteByQueryAction.java]: sure :)

File: server/src/main/java/org/elasticsearch/index/reindex/ScrollableHitSource.java [hunk 2]
```diff
@@ -270,6 +294,14 @@ public BasicHit setRouting(String routing) {
             this.routing = routing;
             return this;
         }
+
+        public void setSeqNo(long seqNo) {
```
[the remark this answers is not in the document]
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/reindex/ScrollableHitSource.java]: That's a good one. It's actually not relevant when indexing from remote (seqnos are only for when update the same index). This class is called "BasicHit" with a java doc of "An implementation of {@linkplain Hit} that uses getters and setters.", so I felt I had to add this, although it's not used. 

Your point actually made me realize I added partial parsing support which wasn't wired and isn't needed. I removed it.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/AsyncDeleteByQueryAction.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ywelsch) [reply modules/reindex/src/main/java/org/elasticsearch/index/reindex/AsyncDeleteByQueryAction.java]: You've changed it in the other class, but not here? 😢

**Review by ywelsch — APPROVED** [review]

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/AsyncDeleteByQueryAction.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (bleskes) [reply modules/reindex/src/main/java/org/elasticsearch/index/reindex/AsyncDeleteByQueryAction.java]: 😢 indeed. fixing.

**Comment by bleskes** [comment]
Thanks @ywelsch !

**Pull request merged**